### ISSUE TITLE
SIGINT sometimes handled twice upon Ctrl+C

### ISSUE THREAD

**Issue by estan** [crossbario/crossbar#278]
Most of the time I get a nice

```
^C2015-03-25 10:32:57+0100 [Controller  31863] Received SIGINT, shutting down.
2015-03-25 10:32:57+0100 [Controller  31863] Main loop terminated.
```

upon stopping the node with Ctrl+C.

But sometimes I get

```
^C2015-03-25 10:32:10+0100 [Controller  31833] Received SIGINT, shutting down.
2015-03-25 10:32:10+0100 [Container   31842] Received SIGINT, shutting down.
2015-03-25 10:32:10+0100 [Container   31842] Connection to node controller lost.
2015-03-25 10:32:10+0100 [Container   31842] Unhandled Error
2015-03-25 10:32:10+0100 [Container   31842] Traceback (most recent call last):
2015-03-25 10:32:10+0100 [Container   31842] Failure: exceptions.RuntimeError: No transport, but disconnect() called.
2015-03-25 10:32:10+0100 [Container   31842] 'While firing onLeave'
2015-03-25 10:32:10+0100 [Controller  31833] Main loop terminated.
```

Notice the *two* `Received SIGINT, shutting down.` messages. It seems the handler for `SIGINT` runs twice for some reason, which I guess is the reason for the subsequent `RuntimeError`. Sometimes this double-handling results in

```
^C2015-03-25 10:31:36+0100 [Controller  31801] Received SIGINT, shutting down.
2015-03-25 10:31:36+0100 [Container   31810] Received SIGINT, shutting down.
2015-03-25 10:31:36+0100 [Container   31810] Connection to node controller lost.
2015-03-25 10:31:36+0100 [Container   31810] Unhandled Error
2015-03-25 10:31:36+0100 [Container   31810] Traceback (most recent call last):
2015-03-25 10:31:36+0100 [Container   31810] Failure: exceptions.RuntimeError: No transport, but disconnect() called.
2015-03-25 10:31:36+0100 [Container   31810] 'While firing onLeave'
2015-03-25 10:31:36+0100 [Container   31810] Unhandled Error
2015-03-25 10:31:36+0100 [Container   31810] Traceback (most recent call last):
2015-03-25 10:31:36+0100 [Container   31810] File "/usr/lib/python2.7/site-packages/twisted/python/context.py", line 118, in callWithContext
2015-03-25 10:31:36+0100 [Container   31810] return self.currentContext().callWithContext(ctx, func, *args, **kw)
2015-03-25 10:31:36+0100 [Container   31810] File "/usr/lib/python2.7/site-packages/twisted/python/context.py", line 81, in callWithContext
2015-03-25 10:31:36+0100 [Container   31810] return func(*args,**kw)
2015-03-25 10:31:36+0100 [Container   31810] File "/usr/lib/python2.7/site-packages/twisted/internet/process.py", line 216, in connectionLost
2015-03-25 10:31:36+0100 [Container   31810] self.proc.childConnectionLost(self.name, reason)
2015-03-25 10:31:36+0100 [Container   31810] File "/usr/lib/python2.7/site-packages/twisted/internet/_posixstdio.py", line 89, in childConnectionLost
2015-03-25 10:31:36+0100 [Container   31810] self.connectionLost(reason)
2015-03-25 10:31:36+0100 [Container   31810] --- <exception caught here> ---
2015-03-25 10:31:36+0100 [Container   31810] File "/usr/lib/python2.7/site-packages/twisted/internet/_posixstdio.py", line 108, in connectionLost
2015-03-25 10:31:36+0100 [Container   31810] protocol.connectionLost(reason)
2015-03-25 10:31:36+0100 [Container   31810] File "/home/estan/Projekt/orexplore/dev/Elvis/full_env/lib/python2.7/site-packages/crossbar/worker/process.py", line 154, in connectionLost
2015-03-25 10:31:36+0100 [Container   31810] sys.exit(1)
2015-03-25 10:31:36+0100 [Container   31810] exceptions.SystemExit: 1
2015-03-25 10:31:36+0100 [Router      31804] Received SIGINT, shutting down.
2015-03-25 10:31:36+0100 [Router      31804] Connection to node controller lost.
2015-03-25 10:31:36+0100 [Router      31804] Unhandled Error
2015-03-25 10:31:36+0100 [Router      31804] Traceback (most recent call last):
2015-03-25 10:31:36+0100 [Router      31804] Failure: exceptions.RuntimeError: No transport, but disconnect() called.
2015-03-25 10:31:36+0100 [Router      31804] 'While firing onLeave'
2015-03-25 10:31:36+0100 [Router      31804] Unhandled Error
2015-03-25 10:31:36+0100 [Router      31804] Traceback (most recent call last):
2015-03-25 10:31:36+0100 [Router      31804] File "/usr/lib/python2.7/site-packages/twisted/python/context.py", line 118, in callWithContext
2015-03-25 10:31:36+0100 [Router      31804] return self.currentContext().callWithContext(ctx, func, *args, **kw)
2015-03-25 10:31:36+0100 [Router      31804] File "/usr/lib/python2.7/site-packages/twisted/python/context.py", line 81, in callWithContext
2015-03-25 10:31:36+0100 [Router      31804] return func(*args,**kw)
2015-03-25 10:31:36+0100 [Router      31804] File "/usr/lib/python2.7/site-packages/twisted/internet/process.py", line 216, in connectionLost
2015-03-25 10:31:36+0100 [Router      31804] self.proc.childConnectionLost(self.name, reason)
2015-03-25 10:31:36+0100 [Router      31804] File "/usr/lib/python2.7/site-packages/twisted/internet/_posixstdio.py", line 89, in childConnectionLost
2015-03-25 10:31:36+0100 [Router      31804] self.connectionLost(reason)
2015-03-25 10:31:36+0100 [Router      31804] --- <exception caught here> ---
2015-03-25 10:31:36+0100 [Router      31804] File "/usr/lib/python2.7/site-packages/twisted/internet/_posixstdio.py", line 108, in connectionLost
2015-03-25 10:31:36+0100 [Router      31804] protocol.connectionLost(reason)
2015-03-25 10:31:36+0100 [Router      31804] File "/home/estan/Projekt/orexplore/dev/Elvis/full_env/lib/python2.7/site-packages/crossbar/worker/process.py", line 154, in connectionLost
2015-03-25 10:31:36+0100 [Router      31804] sys.exit(1)
2015-03-25 10:31:36+0100 [Router      31804] exceptions.SystemExit: 1
2015-03-25 10:31:36+0100 [Router      31804] (TCP Port 8080 Closed)
2015-03-25 10:31:36+0100 [Router      31804] Main loop terminated.
2015-03-25 10:31:36+0100 [Controller  31801] GuestWorkerClientProtocol.connectionLost: [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.error.ConnectionDone'>: Connection was closed cleanly.
        ]
2015-03-25 10:31:36+0100 [Controller  31801] GuestWorkerClientProtocol: INTERNAL ERROR - should not arrive here - [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.error.ConnectionDone'>: Connection was closed cleanly.
        ]
2015-03-25 10:31:36+0100 [Controller  31801] Main loop terminated.
```

But I guess it's a matter of timing whether this occurs or not.

Crossbar should enter its `SIGINT` handler exactly once to avoid this.

**Comment by Paranaix** [comment]
Interessting, this never occured to me. Maybe one can reproduce it by sending SIGINT twice in short occasion with a script?

**Comment by meejah** [comment]
There's actually two (sub-)processes there, each one getting SIGINT. So, should be reproducible by sending the SIGINT to one or the other. Still, the "missing transport" RuntimeErrors shouldn't happen in this case, IMO.

**Comment by estan** [comment]
I should add that I was running a guest worker, in case that has any relevance.

**Comment by oberstet** [comment]
@estan can you still reproduce this using current trunk, since I merged https://github.com/crossbario/crossbar/pull/281 ?

**Comment by estan** [comment]
That looks promising! I'm just about to head home, but will check with trunk when I get in tomorrow.

**Comment by oberstet** [comment]
@estan Thx! These things just _have_ to work smoothly / robustly ..

**Comment by estan** [comment]
OK, back at the office. Unfortunately the problem persist with Git master. Here's a couple of the outputs I got from running `crossbar start` + `Ctrl+C` ~10 times:

```
^C2015-03-31 09:19:38+0200 [Controller  13724] Received SIGINT, shutting down.
2015-03-31 09:19:38+0200 [Router      13727] Received SIGINT, shutting down.
2015-03-31 09:19:38+0200 [Router      13727] Connection to node controller lost.
2015-03-31 09:19:38+0200 [Router      13727] Unhandled Error
2015-03-31 09:19:38+0200 [Router      13727] Traceback (most recent call last):
2015-03-31 09:19:38+0200 [Router      13727] Failure: exceptions.RuntimeError: No transport, but disconnect() called.
2015-03-31 09:19:38+0200 [Router      13727] 'While firing onLeave'
2015-03-31 09:19:38+0200 [Container   13733] Received SIGINT, shutting down.
2015-03-31 09:19:38+0200 [Container   13733] Connection to node controller lost.
2015-03-31 09:19:38+0200 [Container   13733] Unhandled Error
2015-03-31 09:19:38+0200 [Container   13733] Traceback (most recent call last):
2015-03-31 09:19:38+0200 [Container   13733] Failure: exceptions.RuntimeError: No transport, but disconnect() called.
2015-03-31 09:19:38+0200 [Container   13733] 'While firing onLeave'
2015-03-31 09:19:38+0200 [Container   13733] Unhandled Error
2015-03-31 09:19:38+0200 [Container   13733] Traceback (most recent call last):
2015-03-31 09:19:38+0200 [Container   13733] Failure: exceptions.RuntimeError: No transport, but disconnect() called.
2015-03-31 09:19:38+0200 [Container   13733] 'While firing onLeave'
2015-03-31 09:19:38+0200 [Controller  13724] GuestWorkerClientProtocol.connectionLost: [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.error.ConnectionDone'>: Connection was closed cleanly.
        ]
2015-03-31 09:19:38+0200 [Controller  13724] GuestWorkerClientProtocol: INTERNAL ERROR - should not arrive here - [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.error.ConnectionDone'>: Connection was closed cleanly.
        ]
2015-03-31 09:19:38+0200 [Controller  13724] Main loop terminated.
```

and

```
^C2015-03-31 09:20:27+0200 [Controller  13796] Received SIGINT, shutting down.
2015-03-31 09:20:27+0200 [Router      13799] Received SIGINT, shutting down.
2015-03-31 09:20:27+0200 [Router      13799] Connection to node controller lost.
2015-03-31 09:20:27+0200 [Container   13805] Received SIGINT, shutting down.
2015-03-31 09:20:27+0200 [Container   13805] Unhandled Error
2015-03-31 09:20:27+0200 [Container   13805] Traceback (most recent call last):
2015-03-31 09:20:27+0200 [Container   13805] Failure: exceptions.RuntimeError: No transport, but disconnect() called.
2015-03-31 09:20:27+0200 [Container   13805] 'While firing onLeave'
2015-03-31 09:20:27+0200 [Controller  13796] Main loop terminated.
```

I'm running this with the following `.crossbar/config.yml`:

```yaml
controller: {}
workers:

  # WAMP router worker.
  - type: router
    realms:
      - name: orexplore
        roles:
          - name: anonymous
            permissions:
              - uri: '*'
                publish: true
                subscribe: true
                call: true
                register: true
    transports:
      - type: web
        endpoint:
          type: tcp
          port: 8080
        paths:
          /:
            type: static
            directory: ../web
          ws:
            type: websocket

  # Worker for the Magnatek MBG120 x-ray generator simulator.
  - type: guest
    executable: python2
    arguments:
      - orexplore/simulators/mbg120.py
      - --device
      - .crossbar/mbg120_dev
      - --logfile
      - .crossbar/mbg120_simulator.log
    options:
      workdir: ..

  # Worker for the Magnatek MBG120 x-ray generator component.
  - type: container
    options:
      pythonpath: [ .. ]
    components:
      - type: class
        classname: orexplore.components.mbg120.MBG120Component
        realm: orexplore
        transport:
          type: websocket
          endpoint:
            type: tcp
            host: localhost
            port: 8080
          url: ws://localhost:8080/ws
        extra:
          port: mbg120_dev
          baudrate: 9600
```

That is, one guest worker and one container worker.

I haven't looked at the code, but is the controller waiting properly for the workers to die before shutting down itself? I also know that POSIX mandates that SIGINT be sent to the foreground process group (though not sure what that means exactly), perhaps SIGINT is being sent to the workers "behind your back" or something (just shooting in the dark here).

**Comment by estan** [comment]
Hm. Looking at the code in #285, it doesn't look like that is kicking in here? That code delivers SIGTERM to the the worker processes upon reactor shutdown? I would expect that if the shutdown was due to a SIGINT, then SIGINT should be delivered to worker processes (what seems to happen from my outputs). AFAICS it's just that the controller goes to die to early? So perhaps the multiple SIGINTs I'm seeing is really due to normal POSIX behavior, and the controller should just handle this by waiting for the workers to quit (though it is of course possible that the SIGINT was delivered using `kill`, I'm not sure everyone in the group gets the signal then).

**Comment by estan** [comment]
Ah, I see now that there was work in #285 as well. However, shouldn't the `cleanup_worker` method in that patch somehow return a deferred, that is fired once the worker has actually exited (with some reasonable timeout), to prevent the controller reactor from stopping before that happens? I'm new to Twisted so I'm not sure I'm right here :)

**Comment by estan** [comment]
(My last runs above was after the merge of #285 BTW)

**Comment by meejah** [comment]
Yes, good catch @estan we can pause the shutdown by returning a deferred and wait until the worker is done, or timeout (also try KILL on timeout).

I wonder what a "reasonable" amout of time is in this case ;)

**Comment by meejah** [comment]
I'm not sure if this is "the" answer, but I found a problem with shutdown in autobahn, and will submit a PR there after I make a unit-test. Basically, line 966 or so in wamp/protocol.py always calls ``disconnect()`` in ``onLeave``, but if we've already lost our transport the error from ``disconnect`` (that @estan is seeing above) will alway trigger, and basically short-circuit the "nice" shutdown with an error.

I haven't re-created exactly estan's errors, but I can now confirm that (with the above) we always shut down nicely with either "plain" non-WAMP guests or with ApplicationRunner-based guests (that either do or do not successfully join the realm).

**Comment by meejah** [comment]
I believe this should now be closed after the #375 merge, or is there anything still missing from your perspective @estan?

**Comment by estan** [comment]
Absolutely, just an oversight on my part :) Thanks again. Closing.
